refactor: Replace common patterns from various transaction methods with common functions

danieljbruce opened the pull request:
**Summary**

This PR reduces code required in the transaction's get, runQuery, runAggregationQuery and commit methods by replacing similar blocks of code with common functions. Reducing the amount of code in each function reduces technical debt and makes it easier to spot decorator patterns that might simplify code even further.

**Changes**

- Added a function called **#sendUserCallbackData** to increase code reuse. This replaces the custom code contained in each call to withBeginTransaction in get, commit, runQuery and runAggregationQuery. Response data from the resolvers in each of these functions is now provided in an array format so that it can be processed by sendUserCallbackData. This lets us eliminate each of the custom callbacks in calls to withBeginTransaction.
- Added a function called **callbackWithError** which eliminates extra code used to build each resolver. This reduces the amount of code required in get, commit, runQuery and runAggregationQuery used to build a resolver because now a callback in the expected format is provided by this function.
- Finally the prior two changes enforced stronger type alignment so in UserCallbackData the error type was changed so that it couldn't be undefined to solve come compiler errors. This is expected because callbacks either receive errors or null values in the first callback argument throughout the codebase so this provides an improvement to how types line up.

danieljbruce committed:
Change resolvers

Change the resolvers so that they all follow the same pattern.

danieljbruce committed:
Add a wrapWithBeginTransaction function for get

Generalize the calls to withBeginTransaction. Don’t allow the error type to be undefined in UserCallbackData since the error never actually will be undefined. This is necessary to solve various compile time errors.

danieljbruce committed:
Use wrapWithBeginTransaction in more places

commit and runQuery can now use #wrapWithBeginTransaction and this will reduce the amount of code necessary in each function respectively.

danieljbruce committed:
Add callbackWithError function

This function replaces a common pattern that occurs in transactions where we provide a callback that accepts an error as the first argument and some generic type that extends an array of any as the rest of the arguments.

danieljbruce committed:
Add return type to callbackWithError

Adding a return type just makes type checking stronger in the client library. We should introduce it here to make it more likely a compiler error will be thrown if the function is misused.

danieljbruce committed:
Delete code that is not used

When playing around with code to return a standard callback, code was written that is not required anymore. Also, rename StandardCallbackArgumentsAny.

danieljbruce committed:
There is no need to define a separate type

The UserCallbackArguments do not need a separate type if they are only used once. Define the type inline where it is needed.

danieljbruce committed:
TODO no longer applies

danieljbruce committed:
Remove TODO

danieljbruce committed:
Inline the error type

Error or null is only used in two places so inline it here.

danieljbruce committed:
Add a comment for callbackWithError

callbackWithError is an abstract function that needs to be explained. This comment helps with that.

danieljbruce committed:
Added comments to document sendUserCallbackData

sendUserCallbackData needs to be described so that its purpose is clear.

danieljbruce committed:
Move function

Move function to the right order alphabetically

danieljbruce committed:
Do not define a commit type only used once

Do not define a separate variable for only one use.

danieljbruce committed:
Replace the type in the comment with Function

Replace with Function to eliminate references to the generic parameter which make CI tests fail.

danieljbruce committed:
Just use function type for the callback

No need to mention the argument types and return types.

daniel-sanche commented:
I'm a little concerned by how complex this change is getting, and this PR seems to be a step in the wrong direction. There are a lot of concepts being used, and they all interact in ways that's hard to follow.

I tried to trace the stack for a commit, and found
`transaction.commit -> sendUserCallback -> withBeginTransaction -> resolver -> runCommit -> callbackWithError -> PromiseResolveFunction -> ...`

From there, it wasn't completely clear to me who was calling the PromiseResolveFunction and where the arguments came from. There's just too many layers of abstraction and callbacks calling other callbacks. There must be a way to simplify this, right?

 If not, maybe adding a lot more comments would be helpful

daniel-sanche commented:
Some of that feedback applies to PR 1205 as well, but for this one specifically: the change is pretty much line-of-code neutral, and adds two more layers of indirection with `sendUserCallbackData` and `callbackWithError`. So IMO it's probably not worth making this change, unless there are other benefits I'm missing

danieljbruce commented:
I think this PR cleans up a little bit of technical debt. With the way it was before we are applying the same pattern in four places, but writing the pattern out explicitly each time which we shouldn't do. I see how the indirection could be confusing though.

Overall it isn't too critical to get this in because its just an internal change that cleans things up so if you want to leave it for now then perhaps it is fine.

danieljbruce commented:
Regarding the stack trace, without callbackWithError you will get a more anonymous callback which might be even harder to debug. We could also think about combining sendUserCallback and withBeginTransaction into one function that has both code and then it would be easier to follow.

daniel-sanche commented on src/transaction.ts:
```diff
@@ -81,6 +80,31 @@ enum TransactionState {
+/**
+ * This function helps build resolvers from common callback data. Callback data
+ * provided by many functions contains error information in the first argument
+ * and an arbitrary response across all other arguments. It is common to build a
```
Would it be possible to make resolver into an object, and replace this with a constructor?

daniel-sanche commented:
Yeah, IMO this is a step towards more complexity, so probably better to leave it out. I'd look into combining sendUserCallback with withBeginTransaction, and callbackWithError with Resolver. See if you can flatten the call stack. I really don't think it should need this level of complexity

But maybe a node expert would have different thoughts too, so maybe a different reviewer would have different thoughts?

danieljbruce committed:
Remove withBeginTransaction

withBeginTransaction is only called by one function so remove it to condense call stack.

danieljbruce commented:
Closing this PR in favour of https://github.com/googleapis/nodejs-datastore/pull/1212.
